#4973 Implement Edit Profile page with shared Request Access logic

lhawkman27 opened the pull request:
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #4973 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

_**NOTE: when testing ACCESS REQUEST make sure you start from a initial access request state**_

**Test Edit Profile UI functionality**
1. Open http://localhost:3000/ and sign in.
2. Navigate to Profile page (click Profile tab)
3. View the new Profile page layout
<img width="1524" height="812" alt="Screenshot 2025-08-14 at 10 32 21 AM" src="https://github.com/user-attachments/assets/8bb55d5b-2df3-4fa5-b067-b87e71140201" />

5. Verify you see Name, User Type, Jurisdiction Type, State, Territory, or Tribe, and FRA radio select fields with respective values. (NOTE: if you have Regional Staff role you will see fields such as Name, User Type, and Regional Office(s))
7. Click Edit Profile Information button
8. Verify Edit Profile state happens (changes from read-only to edit) and page changes
<img width="1511" height="855" alt="Screenshot 2025-08-14 at 10 34 55 AM" src="https://github.com/user-attachments/assets/39ba2a2b-8ef5-4e89-a4a4-fd703f9ecb54" />

10. Click the Cancel button and verify page goes back into **Read-Only Mode**
11. Click Edit Profile Information button
12. Try and hit the Save Change button
13. Verify you see an error alert ppear that says _No Change have been made_. 
<img width="1587" height="834" alt="Screenshot 2025-08-13 at 4 03 45 PM" src="https://github.com/user-attachments/assets/3855c8ae-0d37-4355-8fff-5ba9e17b521b" />

15. Now select a field or fields and backspace out the value and click away. 
16. Verify you see an error alert that says _There are [#Num of fields outlined in red] errors in form_. (fields errors should handle the same way as they did previously the users would create Access Request)
<img width="1456" height="860" alt="Screenshot 2025-08-13 at 4 04 15 PM" src="https://github.com/user-attachments/assets/8b753a49-5f0f-47bb-9d60-8a6744a2fc70" />

18. Now enter back in values for the field or field make sure one is at least different from its previous state. 
19. Submit change (**click Save Changes button**) and you should see page go back to read-only profile layout 

**Test Request Access Edit Ui Functionality**
1. Open http://localhost:3000/ and sign in.
2. Navigate to Admin page (DAC)
3. Scroll down to Users and select Users
4. Select your User account
5. Update your Account Approval Status to **PENDING** (_make sure if you have a STT associated with your account you want to remove it before you save the update --> Remove STT --> Save and Continue Editing --> Update account approval status to Pending --> Save_)
6. Navigate back to Home page
7. Verify you see Request Submitted page with new layout (like you would when you first are setting up your TDP/TANF account)
8. Verify the Pending approval alert message is showing as well.
<img width="1684" height="771" alt="Screenshot 2025-08-14 at 10 38 57 AM" src="https://github.com/user-attachments/assets/4cd807b0-7afe-47a0-9470-9bcb2276d9f0" />

10. Verify you see a new button present for editing the Access Request (shown in image above)
11. Click Edit Access Request
12. Verify Edit Request state happens (changes from read-only to edit) and page changes
<img width="1390" height="866" alt="Screenshot 2025-08-14 at 10 39 43 AM" src="https://github.com/user-attachments/assets/12d783da-680d-4228-b510-3ae1baa4d98e" />

15. Click the Cancel button and verify page goes back into **Read-Only Mode**
16. Click Edit Access Request
17. **The error handling works the same as it does when Editing Profile** (and form errors are handled the same as they were before when creating a Access Request) --> _Repeat steps 10 - 14 if you want to test error states_
<img width="1362" height="848" alt="Screenshot 2025-08-14 at 10 40 57 AM" src="https://github.com/user-attachments/assets/f0126b21-f341-4812-b843-911c372f22a8" />
<img width="1342" height="867" alt="Screenshot 2025-08-14 at 10 41 09 AM" src="https://github.com/user-attachments/assets/1faa49c8-e969-4ce8-92da-df36d4c1d6e6" />

19.  Update a field and Submit Change (**Click Update Request Button**)
20. Verify page goes back to read-only Access Request Submitted layout 

**Test Regional Staff Role Edit Ui Functionality (the UI is a bit different)**

PROFILE --------------------------------------
1. Open http://localhost:3000/ and sign in.
2. Navigate to Admin page (DAC)
3. Scroll down to Users and select Users
4. Select your User account
5. Update your role to **OFA Regional Staff** (_make sure if you have a STT associated with your account you want to remove it before you save the update --> Remove STT --> Save and Continue Editing --> Update Group role  and change email so it ends with @acf.hhs.gov --> Save_)
6. Navigate to Profile page (click Profile tab)
7. View the new Profile page layout
<img width="1676" height="802" alt="Screenshot 2025-08-13 at 9 09 15 AM" src="https://github.com/user-attachments/assets/c96c2e57-d7d4-4414-94c1-8e8cdeb495d0" />

9. Click Edit Profile Information button
10. Verify Edit Profile state happens (changes from read-only to edit) and page changes
<img width="1705" height="931" alt="Screenshot 2025-08-13 at 9 07 56 AM" src="https://github.com/user-attachments/assets/5d323a9c-cbf0-47e4-b4b4-fd6133c5562f" />
<img width="1431" height="881" alt="Screenshot 2025-08-13 at 9 08 03 AM" src="https://github.com/user-attachments/assets/32624e92-4e61-466a-bb18-36710a65a7d3" />

12.  **The error handling works the same as it does when Editing Profile** (and form errors are handled the same as they were before when creating a Access Request) --> _Repeat steps 10 - 14 from above if you want to test error states_
<img width="1355" height="773" alt="Screenshot 2025-08-14 at 12 42 05 PM" src="https://github.com/user-attachments/assets/e30c3349-d84c-4e13-951e-d98504fabfd1" />

14. Since you are a Regional Staff role if you uncheck all your regions, you should get an error state for that now as well
<img width="1346" height="764" alt="Screenshot 2025-08-14 at 12 40 45 PM" src="https://github.com/user-attachments/assets/045dabc2-65d2-4a6b-890a-2acb0df3379e" />

15. Continue with update the fields and get form back into valid state
16. Submit and page should go back into Read-Only Mode

ACCESS REQUEST -------------------------------------------------------
1. Navigate back to Admin (DAC)
2. Update your account approval status to **PENDING**
3. Navigate back to Home page
4. Verify you see Request Submitted page with new layout (like you would when you first are setting up your TDP/TANF account)
<img width="1715" height="935" alt="Screenshot 2025-08-13 at 9 06 08 AM" src="https://github.com/user-attachments/assets/f17f06f2-c85a-40ea-b466-562da3bd80e0" />

6. You can follow the steps above for editing the Profile and error checking for the Access Request
<img width="1564" height="826" alt="Screenshot 2025-08-13 at 9 06 56 AM" src="https://github.com/user-attachments/assets/46fd3939-dbd1-4117-969b-2e24adc6dd9e" />
<img width="1445" height="806" alt="Screenshot 2025-08-14 at 12 42 33 PM" src="https://github.com/user-attachments/assets/f77c44ec-7065-4930-91f1-5d970d2a2726" />

8. Update the fields to get form into valid update submission state
9. Submit and page should go back into Read-Only Mode

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ]  Request Access form logic is refactored for reuse between onboarding and Edit Profile flows
+ [ ] Edit Profile page implemented per [Figma Spec](https://www.figma.com/design/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=11302-9503&t=hFDt4oAxPEPANwCn-0)
+ [ ] Form validation matches specified logic in the design
+ [ ] Stub PATCH endpoints are created if backend update work is not yet available
+ [ ] Frontend tests written for form state and mock PATCH requests
+ [ ] Testing Checklist has been run and all tests pass
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

Commit:
#4973 started implementing a new component for updated user access request info, may not need this but want to flesh out the UI and then if i can refactor I will

Commit:
#4973 more sizing and placement work for ui

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

Commit:
#4973 created/updated UI components abd updated error message logic

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

Commit:
#4973 updating some UI styling

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

Commit:
#4973 need to add some check for showing certain info for access request

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

Commit:
#4973 updating some data passed to edit form

Commit:
#4973 adding editing logic for Profile UI and added some tests for RequestForm and UserAccessInfo; still needing to clean up the UI and try to get styling working properly

Commit:
#4973 made a good bit of refactoring changes to the codebase to make things cleaner and more reusable

Commit:
#4973 refactored code some more and added some new jest tests for new components. Need to still fix breaking layout white space issue and finish fixing unit tests

Commit:
#4973 fixing test

Commit:
#4973 updated some more failing tests

Commit:
#4973 finally got all tests passing and code coverage taken care of. Ui tweaks as well. Need to do a little more clean up

Commit:
#4973 finished cleaning up code

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

Commit:
#4973 last minute updates for mocking the api call while add in the actual call once it's merged into develop

Commit:
#4793 last minute updates and clean up

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

lhawkman27 commented:
**_Planning to add the backend API call in once that is merged into develop_**

Commit:
#4973 fixing some test

Commit:
#4973 fixed Profile tests

Commit:
#4973 fixing some mocks for test

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

elipe17 reviewed (changes requested):
Took a quick first pass at testing this and ran into the following bugs/questions

- Upon signing in and requesting access with FRA as standard STT, when i go to the edit page the FRA radio is toggled to "No" even though the DAC states I have selected FRA. Could cause a lot of users to accidentally disable FRA access unknowingly.
- Upon approving my STT user who requested FRA, I am viewing the profile page and it is indicating I do not have FRA access
- On the Request Submitted page I don't think we should show the "User Type" field. When initially requesting access this will always be null
- Bug: When requesting access as a regional user I am getting a 400 response with: `{"regions":["Incorrect type. Expected pk value, received dict."]}`
- Requesting access as an ACF user with no regions shows blank "User Type" and blank "Regions" fields on the Request Submitted page. Should we just exclude those in this case?
- When approved as an ACF user with no regions, I see the blank "Regions" field on the profile page
- Bug: As an approved ACF user with OFA Admin role, go to profile page, edit profile, immediately save changes to see form error, select yes to choose regions, save changes again to see regions form error and the "no changes" form error, select a region to clear the region form error, verify the "no changes" form error persists and you cannot save and that it is only cleared when updating the first/last name form fields

Commit:
#4973 made some updates based off code review

Commit:
#4973 fixes for unit test due to field removal some refactoring

Commit:
#4973 unit test fixes

Commit:
#4973 updated unit tests and fixed some validation logic when editing to fix bug issues

lhawkman27 commented:
Made Ui changes discussed in Dev Office Hours with bug fixes

- Removed User Type field from showing at all
- Added None selected as a value for an ACF user that hasn't selected any regions on access request creation (if they edit and add a region that region will show up in place of the None Selected)
- IF a ACF user with no Regions selected views their profile they will only see their name until they choose to edit
- Fixed bugs found with region selection and validation logic
- Fixed request submission issue response `400 {"regions":["Incorrect type. Expected pk value, received dict."]}`
- Refactored code to make it a bit more readable and clean

Commit:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

Commit:
#4973 fixed broken unit tests

raftmsohani reviewed (approved):
Tested locally, works as advertised

Commit:
#4973 fixed test issue and refactored the code a bit more

elipe17 commented on tdrs-frontend/src/components/Home/Home.jsx:
```diff
@@ -1,42 +1,41 @@
   const userAccessInReview = useSelector(accountIsInReview)
   const userAccessRequestApproved = useSelector(accountStatusIsApproved)
 
+  const [isEditing, setIsEditing] = useState(false)
```
Why do we need two "Editing" state setters? Could this be accomplished with one?

elipe17 commented on tdrs-frontend/src/components/Profile/JurisdictionLocationInfo.jsx:
```diff
@@ -0,0 +1,24 @@
+}) => {
+  let label = 'State'
+  if (jurisdictionType === 'territory') label = 'Territory'
+  else if (jurisdictionType === 'tribe') label = 'Tribe'
```
Could we make these Enums to avoid accidental incorrect string usages?

elipe17 replied:
Same thing with formType?

elipe17 reviewed (approved):
Nice job! This was a lot of work and a lot states to test! Will you write a two follow on tickets. First, one to add E2E tests for the edit profile flow. Second, could you write up a ticket to allow the user to update their STT only while in the request access state?

Commit:
#4973 fixed up test coverage issue

Commit:
#4973 added enums in the string values

Commit:
#4973 fixed import issue

mattcoleanderson committed:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

kristen-shaw4 commented:
Can we add an alert that an edit request was successfully submitted & what users can expect next? I submitted a change during my testing, and it's not clear to me who approves the change or how long it will take, and when I go back to edit all of my previous info is still there, so it's confusing from a user perspective whether my changes went through or not.

There's a [Figma design](https://www.figma.com/design/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=11341-6719&t=yKpH7VXiniuDrXGs-1) that is used for an access request that we can use for the edit request.

Do we also send an email confirming that the edit request was submitted and then when it's approved?

@elipe17 @mattcoleanderson @lhuxraft

mattcoleanderson committed:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

mattcoleanderson commented:
> Can we add an alert that an edit request was successfully submitted & what users can expect next? I submitted a change during my testing, and it's not clear to me who approves the change or how long it will take, and when I go back to edit all of my previous info is still there, so it's confusing from a user perspective whether my changes went through or not.
> 
> There's a [Figma design](https://www.figma.com/design/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=11341-6719&t=yKpH7VXiniuDrXGs-1) that is used for an access request that we can use for the edit request.
> 
> Do we also send an email confirming that the edit request was submitted and then when it's approved?

@kristen-shaw4 This ticket is for implementing the the frontend framework that will allow for profile editing. But we have another ticket, [4974](https://github.com/raft-tech/TANF-app/issues/4974), for implementing the backend logic. So as of right now, if you were to submit an edit profile request you won't see any change, alert, or email confirmation.

mattcoleanderson committed:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

mattcoleanderson committed:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

mattcoleanderson committed:
add a reducer for updateUserRequest and implement the new backend endpoint

mattcoleanderson committed:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

mattcoleanderson committed:
guard against null region on update user profile request

mattcoleanderson committed:
rename mockUpdateUserRequest.js to updateUserRequest.js

mattcoleanderson committed:
add pending_requests to the UserProfileSerializer so it is available to show the pending requests banner on load of the profile page

mattcoleanderson committed:
add email support for change requests

mattcoleanderson committed:
make regions and has_fra_access humand readable in email

mattcoleanderson committed:
add rejected template and put current and new values in unordered list

mattcoleanderson committed:
fix backend linting issues

mattcoleanderson committed:
fix frontend tests

mattcoleanderson commented:
@elipe17 @raftmsohani @jtimpe Hey guys, finished up integrating the backend additions from @raftmsohani into this and am putting this back into code review agains since I've made various significant changes and would appreciate if y'all could review it. The following are some changes I've made (including a few backend changes to align with the needs of the frontend):

- implement a redux action/reducer/selector to call the PATCH users/update_profile endpoint on profile change submission
- add `pending_requests` field to the `UserProfileSerializer` model to show a banner on the Profile page if the user has any pending change requests
- Add email support to the ChangeRequest model so an email is sent on approval or rejection of a change request
    - email will have the field approved/rejected, the new and requested field values, and any notes from the admin who reviewed the request

mattcoleanderson committed:
add test for reducer/updateUserRequest

mattcoleanderson committed:
add full coverage for actions/updateUserRequest.js

mattcoleanderson committed:
fix linting issues

jtimpe reviewed (approved):
Works great!

raftmsohani commented:
Frontend is sending a request to change has_fra_access although I only changed my name:

<img width="1385" height="302" alt="Screenshot 2025-09-26 at 11 27 01 AM" src="https://github.com/user-attachments/assets/a8207fb4-0922-4187-8ed6-61ccf66f61d6" />

raftmsohani commented on tdrs-backend/tdpservice/email/helpers/profile_change_request.py:
```diff
@@ -0,0 +1,93 @@
+
+def send_change_request_status_email(change_request, isApproved: bool, url):
+    """Send an email to a user when their profile change request is approved/denied."""
+    from tdpservice.users.models import User
```
Did you have to import here?

mattcoleanderson replied:
Was there for a type hint. not sure why it ended up under the method. I removed it

elipe17 reviewed (changes requested):
I ran into a myriad of bugs, see below:

- Started without FRA access, requested an edit to have FRA and the DAC showed incorrect state.
<img width="1025" height="111" alt="Screenshot 2025-09-26 at 10 22 37 AM" src="https://github.com/user-attachments/assets/a8f795a9-01f5-4488-b6ea-0ca95abc0789" />

- Non-regional staff ACF users aren't able to request FRA (e.g. DIGIT). Is this expected, should DIGIT not be able to see FRA/request it?
- Regional staff bug in DAC. I think this is coming from the PR Mo recently merged. 
https://github.com/user-attachments/assets/b5571711-35e0-4795-b2cf-5b3a176e575f

- If an ACF user starts out without regions, requests access to regions, and has them approved, the DAC is locked for that user because they will be a different group than OFA Regional Staff so we won't ever be able to change their profile from the DAC because of the checks
- As an approved ACF user with regions selected. If I request to remove all regions no change request is sent and I always have regions

mattcoleanderson committed:
make existing_permission a bool

mattcoleanderson committed:
update request_change method on change request model to grab user permission directly so current_value is correct for superusers as well

mattcoleanderson commented:
> Frontend is sending a request to change has_fra_access although I only changed my name:
> 
> <img alt="Screenshot 2025-09-26 at 11 27 01 AM" width="1385" height="302" src="https://private-user-images.githubusercontent.com/97037188/494482099-a8207fb4-0922-4187-8ed6-61ccf66f61d6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg5MTIzNTQsIm5iZiI6MTc1ODkxMjA1NCwicGF0aCI6Ii85NzAzNzE4OC80OTQ0ODIwOTktYTgyMDdmYjQtMDkyMi00MTg3LThlZDYtNjFjY2Y2NmY2MWQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI2VDE4NDA1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyMWE0YmMzNWU0ZGZlZWU3NGQ1NjhiOTFmNDljZGY0MDFkY2JhYzFmOTBmZTUzYjIyMTZlNjU4OGFkMjAyZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AC8atPZBEc0tc05xiPhhdihrx32jQiOnNc6bMZHlpI0">

@raftmsohani this should be fixed now

mattcoleanderson committed:
remove import statement

mattcoleanderson commented:
> I ran into a myriad of bugs, see below:
> 
> * Started without FRA access, requested an edit to have FRA and the DAC showed incorrect state.
> 
> <img alt="Screenshot 2025-09-26 at 10 22 37 AM" width="1025" height="111" src="https://private-user-images.githubusercontent.com/125676261/494495052-a8f795a9-01f5-4488-b6ea-0ca95abc0789.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg5MTIzNTQsIm5iZiI6MTc1ODkxMjA1NCwicGF0aCI6Ii8xMjU2NzYyNjEvNDk0NDk1MDUyLWE4Zjc5NWE5LTAxZjUtNDQ4OC1iNmVhLTBjYTk1YWJjMDc4OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyNlQxODQwNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYjczNmIwZGI3MzUwOTA5NTI5OWRlNjk1MzQwNGVhYmRiNmJkMmY1OGM1ZWE4NzhjNWJlYWYzM2EwMmEzZGJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.AsrvOx6eKIjiBtFjpgvY_EFqhJFxUw3EWQQNxsifi8A">
> * Non-regional staff ACF users aren't able to request FRA (e.g. DIGIT). Is this expected, should DIGIT not be able to see FRA/request it?
> * Regional staff bug in DAC. I think this is coming from the PR Mo recently merged.
>   https://github.com/user-attachments/assets/b5571711-35e0-4795-b2cf-5b3a176e575f
> * If an ACF user starts out without regions, requests access to regions, and has them approved, the DAC is locked for that user because they will be a different group than OFA Regional Staff so we won't ever be able to change their profile from the DAC because of the checks
> * As an approved ACF user with regions selected. If I request to remove all regions no change request is sent and I always have regions

@elipe17 we discussed these in Office hours. Below is a summary of our conclusion:

> * Non-regional staff ACF users aren't able to request FRA (e.g. DIGIT). Is this expected, should DIGIT not be able to see FRA/request it?

This bug only showed up for superusers and should be fixed now

> * Regional staff bug in DAC. I think this is coming from the PR Mo recently merged.
>   https://github.com/user-attachments/assets/b5571711-35e0-4795-b2cf-5b3a176e575f
> * If an ACF user starts out without regions, requests access to regions, and has them approved, the DAC is locked for that user because they will be a different group than OFA Regional Staff so we won't ever be able to change their profile from the DAC because of the checks

Both of these are a result of how Django handles M2M fields on save. Specifically model fields are saved and validated before M2M fields are leading to weird validation issues/deadlocks. @raftmsohani will fix these in his bugfix that is solving a similar issue already: https://github.com/raft-tech/TANF-app/issues/5368

> * As an approved ACF user with regions selected. If I request to remove all regions no change request is sent and I always have regions

The UserProfileChangeRequestSerializer validates regions to be nonempty should it be present on update. We need to ask @ADPennington if there is every a usecase for an ACF user to remove all their currently assigned regions

mattcoleanderson commented:
> * As an approved ACF user with regions selected. If I request to remove all regions no change request is sent and I always have regions

@elipe17 I've spoke with @ADPennington about this and she says there isn't a need for an ACF user to remove their region selection. The only time this might occur would be if they were changing departments, which their account would be deactivated when that happens anyway so this validation makes sense.

mattcoleanderson committed:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic

elipe17 commented:
> > * As an approved ACF user with regions selected. If I request to remove all regions no change request is sent and I always have regions
> 
> @elipe17 I've spoke with @ADPennington about this and she says there isn't a need for an ACF user to remove their region selection. The only time this might occur would be if they were changing departments, which their account would be deactivated when that happens anyway so this validation makes sense.

In this case, can we disallow a regional user from selecting "No" on the edit form to have a better experience? Otherwise I am good.

elipe17 reviewed (approved):
Left one more question for UX, but otherwise I am good. LGTM

victoriaatraft commented:
> Left one more question for UX, but otherwise I am good. LGTM

Per DM with @elipe17 holding off on this to see what see what regional users end up doing

victoriaatraft commented:
<img width="1728" height="962" alt="Screenshot 2025-10-06 at 11 54 27 AM" src="https://github.com/user-attachments/assets/4d415a8b-75f6-4598-ae2c-4fbdcad413dd" />
<img width="1728" height="960" alt="Screenshot 2025-10-06 at 11 43 49 AM" src="https://github.com/user-attachments/assets/4bd96329-3a45-4d29-9b8b-a586057d5da4" />
<img width="1728" height="962" alt="Screenshot 2025-10-06 at 11 43 34 AM" src="https://github.com/user-attachments/assets/04df0b16-dd8e-40d8-9f6d-f16fb007746f" />
<img width="1728" height="958" alt="Screenshot 2025-10-08 at 1 43 05 PM" src="https://github.com/user-attachments/assets/d442b2f8-6935-454b-af2b-cc1dc0f037d3" />
![Uploading Screenshot 2025-10-08 at 1.43.28 PM.png…]()

mattcoleanderson committed:
disable OFA Regional Office radio button in profile edit if user is approved and has an assigned region

mattcoleanderson commented:
> > Left one more question for UX, but otherwise I am good. LGTM
> 
> Per DM with @elipe17 holding off on this to see what see what regional users end up doing

@victoriaatraft You can continue now! Just pushed the change, @elipe17 requested, to disable the `Do you work for an OFA Regional Office?` radio buttons for approved ACF users who have a currently assigned region:

<img width="610" height="818" alt="image" src="https://github.com/user-attachments/assets/fe63b28d-b0f3-4794-8e58-0a5b6ce8c47e" />

mattcoleanderson committed:
test for fieldset disable

reitermb commented:
@mattcoleanderson I'm curious whether it's a light lift to programmatically treat clicking save without having made changes as the same action as cancelling? Just a nice to have so that we're not popping an error message that amounts to "click the other button". 

If that's heavy though I don't know that it's worth significant additional scope in this PR.

mattcoleanderson commented:
> @mattcoleanderson I'm curious whether it's a light lift to programmatically treat clicking save without having made changes as the same action as cancelling? Just a nice to have so that we're not popping an error message that amounts to "click the other button".
> 
> If that's heavy though I don't know that it's worth significant additional scope in this PR.

@reitermb I feel though that if someone thought they had made a change, hit save, and didn't get an error they'd be likely to assume that there change request was successfully submitted. To me it makes more sense to give them an error so they are aware that they attempted to submit a change request but didn't change anything.

victoriaatraft commented:
> > @mattcoleanderson I'm curious whether it's a light lift to programmatically treat clicking save without having made changes as the same action as cancelling? Just a nice to have so that we're not popping an error message that amounts to "click the other button".
> > If that's heavy though I don't know that it's worth significant additional scope in this PR.
> 
> @reitermb I feel though that if someone thought they had made a change, hit save, and didn't get an error they'd be likely to assume that there change request was successfully submitted. To me it makes more sense to give them an error so they are aware that they attempted to submit a change request but didn't change anything.

+1 @mattcoleanderson

victoriaatraft reviewed (approved):
LGTM

mattcoleanderson committed:
Merge branch 'develop' into 4973-implement-edit-profile-page-with-shared-request-access-logic
